File: web/blocks/attributes/attribute.dart
/*
 * NetTango
 * Copyright (c) 2020 Michael S. Horn, Uri Wilensky, and Corey Brady
 *
 * Northwestern University
 * 2120 Campus Drive
 * Evanston, IL 60613
 * http://tidal.northwestern.edu
 * http://ccl.northwestern.edu

 * This project was funded in part by the National Science Foundation.
 * Any opinions, findings and conclusions or recommendations expressed in this
 * material are those of the author(s) and do not necessarily reflect the views
 * of the National Science Foundation (NSF).
 */

part of NetTango;

/// Represents the paramter or property options for a block
abstract class Attribute {

  /// parameter id - unique per block
  int id;

  get uniqueId => "${block.workspace.containerId}-${block.id}-${id}";

  /// parent block
  Block block;

  /// parameter type: can be "bool", "num", "int", "range", "text", or "select"
  String get type;

  /// name of the parameter (e.g. degrees, length)
  String name = "";

  /// short unit name that will be displayed after the value (e.g. %, px, m, mm, s)
  String unit = "";

  String getValue();
  void setValue(String valueString);

  String getDefaultValue();
  void setDefaultValue(String defaultString);

  String getDisplayValue() => "${this.getValue()}${this.unit}";

  Attribute(this.block, this.id) {
    if (this.id != null) {
      if (id >= this.block.nextParamId) {
        this.block.nextParamId = id + 1;
      }
    } else {
      id = this.block.nextParamId++;
    };
  }

  Attribute.clone(this.block, Attribute source, bool isSlotBlock) {
    this.id = source.id;
    this.name = source.name;
    this.unit = source.unit;
  }

  // just so we can clone without knowing the type
  Attribute clone(Block block, bool isSlotBlock);

  // should the value for this attribute be quoted in code?
  bool shouldQuote();

  // parameters are meant to display inline with just a value
  DivElement drawParameter() {
    DivElement paramDiv = new DivElement();
    paramDiv.innerText = this.getDisplayValue();
    paramDiv.classes.add("nt-attribute-value");
    paramDiv.classes.add("${block.getStyleClass()}-attribute");
    if (block.blockColor != null) { paramDiv.style.color = block.blockColor; }
    if (block.textColor != null) { paramDiv.style.backgroundColor = block.textColor; }

    final updateValue = () { paramDiv.innerText = this.getDisplayValue(); };
    paramDiv.onClick.listen( (event) {
      final target = (event.target as DivElement);
      final parent = target.offsetParent;
      final left = parent.offsetLeft + target.offsetLeft + event.offset.x;
      final top  = parent.offsetTop  + target.offsetTop  + event.offset.y;
      showParameterDialog(left.floor(), top.floor(), updateValue);
    });
    return paramDiv;
  }

  // properties display stand-alone with an identifier
  DivElement drawProperty() {
    DivElement propDiv = new DivElement();
    propDiv.classes.add("nt-property");
    DivElement propName = new DivElement();
    propName.classes.add("nt-property-name");
    propName.innerText = "\u2022 $name";
    propDiv.append(propName);
    propDiv.append(drawParameter());
    return propDiv;
  }

  showParameterDialog(int x, int y, Function acceptCallback);
}

File: src/blocks/attributes/attribute.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { Attribute, AttributeValue } from "../../types/types"
import { BlockInstanceUI } from "../block-instance"

type AttributeTypes = "bool" | "num" | "int" | "range" | "text" | "select"

/// Represents the paramter or property options for a block
abstract class AttributeUI {

  readonly def: Attribute
  readonly a: AttributeValue
  readonly block: BlockInstanceUI

  get uniqueId(): string { return `${this.block.workspace.containerId}-${this.block.def.id}-${this.def.id}` }

  getDisplayValue(): string { return `${this.a.value}${this.def.unit}` }

  constructor(def: Attribute, a: AttributeValue, block: BlockInstanceUI) {
    this.def = def
    this.a = a
    this.block = block
  }

  // parameters are meant to display inline with just a value
  drawParameter(): HTMLDivElement {
    const paramDiv = document.createElement("div")
    const updateValue = () => { paramDiv.innerText = this.getDisplayValue(); }
    updateValue()
    paramDiv.classList.add("nt-attribute-value")
    paramDiv.classList.add(`${this.block.getStyleClass()}-attribute`)
    if (this.block.def.blockColor !== null) { paramDiv.style.color = this.block.def.blockColor; }
    if (this.block.def.textColor !== null) { paramDiv.style.backgroundColor = this.block.def.textColor; }

    paramDiv.addEventListener("click", (event) => {
      const target = (event.target as HTMLDivElement)
      const parent = (target.offsetParent as HTMLDivElement)
      const left = parent.offsetLeft + target.offsetLeft + event.offsetX
      const top  = parent.offsetTop  + target.offsetTop  + event.offsetY
      this.showParameterDialog(Math.floor(left), Math.floor(top), updateValue)
    })
    return paramDiv
  }

  // properties display stand-alone with an identifier
  drawProperty(): HTMLDivElement {
    const propDiv = document.createElement("div")
    propDiv.classList.add("nt-property")
    const propName = document.createElement("div")
    propName.classList.add("nt-property-name")
    propName.innerText = `\u2022 ${this.def.name}`
    propDiv.append(propName)
    propDiv.append(this.drawParameter())
    return propDiv
  }

  abstract showParameterDialog(x: number, y: number, acceptCallback: () => void): void
}

export { AttributeUI, AttributeTypes }
